lib/src/productsearch_client.dart

```diff
@@ -0,0 +1,154 @@
+  static const int _defaultTimeout = 15000;
+
+  static const String _lastReqidKey = 'visenze_query_id_';
+
+  final String _appKey;
+  final String _placementId;
+  final VisenzeTracker _tracker;
+  late final int _timeoutInMs;
+  late final bool _useStaging;
+  late final SharedPreferences _prefs;
+
+  static Future<ProductSearchClient> create(
+      String appKey, String placementId, VisenzeTracker tracker,
+      {bool? useStaging, int? timeout}) async {
+    var client = ProductSearchClient(appKey, placementId, tracker,
+        useStaging: useStaging, timeout: timeout);
+    await client._init();
+    return client;
+  }
+
+  ProductSearchClient(this._appKey, this._placementId, this._tracker,
+      {bool? useStaging, int? timeout}) {
+    _timeoutInMs = timeout ?? _defaultTimeout;
+    _useStaging = useStaging ?? false;
+  }
+
+  Future<http.Response> imageSearch(Map<String, dynamic> searchParams) async {
+    Map<String, dynamic> params = getCommonParams();
+    params.addAll(_getAuthParams());
+    params.addAll(searchParams);
+
+    http.Response response;
+
+    if (params['image'] != null) {
+      dynamic file = params['image'];
+      params.remove('image');
+      params = params.map((key, value) => MapEntry(key, value.toString()));
+      Uri uri = Uri.https(_getEndpoint(), _pathSearch, params);
+      response = await _uploadPost(uri, file);
+    } else {
+      params = params.map((key, value) => MapEntry(key, value.toString()));
+      Uri uri = Uri.https(_getEndpoint(), _pathSearch, params);
+      response = await _post(uri);
+    }
+    _onSearchCompleted(response);
+    return response;
+  }
+
+  Future<http.Response> idSearch(
+      String pid, Map<String, dynamic>? requestParams) async {
+    Map<String, dynamic> params = getCommonParams();
+    params.addAll(_getAuthParams());
+    params.addAll(requestParams ?? {});
+    params = params.map((key, value) => MapEntry(key, value.toString()));
+    Uri uri = Uri.https(_getEndpoint(), '$_pathRec/$pid', params);
+    var response = await _get(uri);
+    _onSearchCompleted(response, pid);
+    return response;
+  }
+
+  Map<String, dynamic> getCommonParams() {
+    return {
+      'ts': DateTime.now().millisecondsSinceEpoch,
+      'va_sdk': _sdk,
+      'va_sdk_version': _version,
+      'va_uid': _tracker.userId,
+      'va_sid': _tracker.sessionId
+    };
+  }
+
+  String? get lastSuccessQueryId {
+    return _prefs.getString('$_lastReqidKey$_placementId');
+  }
+
+  bool _isResponseSuccess(http.Response response) {
+    final resp = jsonDecode(response.body);
+    return response.statusCode == 200 &&
+        ((resp['result'] != null && resp['result'].length > 0) ||
+            (resp['objects'] != null && resp['objects'].length > 0));
+  }
+
+  /// Send a result load
+  void _onSearchCompleted(http.Response response, [String? pid]) {
+    if (_isResponseSuccess(response)) {
+      final resp = jsonDecode(response.body);
+      _sendResultLoadEvent(resp['reqid'], pid);
```

Comment:
need to make sure that there is at least 1 result in the response before sending the result load event

Reply:
update